incubator/hnc/pkg/reconcilers/hnc_config.go

```diff
@@ -136,28 +145,45 @@ func (r *ConfigReconciler) createObjectReconcilers(inst *api.HNCConfiguration) {
 	// from the forest for the object reconciliation, after we create ObjectReconcilers but before we write the
 	// ObjectReconcilers to the types list. As a result, objects of the newly added types might not be propagated correctly
 	// using the latest forest structure.
+	//
+	// For syncing an object reconciler, the mutex is guarding the read access to the `namespaces` field in the forest. The
+	// `namespaces` field is a shared resource between various reconcilers.
 	r.Forest.Lock()
 	defer r.Forest.Unlock()
 
 	for _, t := range inst.Spec.Types {
 		gvk := schema.FromAPIVersionAndKind(t.APIVersion, t.Kind)
-		if r.Forest.HasTypeSyncer(gvk) {
-			continue
+		if ts := r.Forest.GetTypeSyncer(gvk); ts != nil {
+			err := r.syncObjectReconciler(ctx, t.Mode, ts)
+			// Return sync error here so that we can retry the reconciliation.
+			if err != nil {
+				return err
+			}
```

Comment:
For future reference, this is a good time to combine the lines:

```
if err := r.syncObjectReconciler(ctx, t.Mode, ts); err != nil {
  return err // retry the reconciliation
}
```

But I won't hold up the merge for this :)

Reply:
Also, why not fold the two-line function below into this one?

```
if err := ts.SetMode(ctx, r.Log, t.Mode); err != nil {
  return err // retry the reconciliation
}
```

You can move the logging into the method.

Reply:
Thanks for the suggestion! Updated to use the second example.